src/main/java/com/elytradev/concrete/resgen/ConcreteResourcePack.java

```diff
@@ -0,0 +1,328 @@
+		}
+
+		if (realPack == null || this.realResourcePack == null) {
+			throw new MissingRealpackException(modID);
+		}
+
+		//Add our pack as a default pack, use FMLClientHandler's field for this so we don't need to worry about obf names
+		resourcePackList.get(FMLClientHandler.instance()).add(this);
+
+		// Confirms that our resourcepack is available as soon as possible to prevent missing resource errors.
+		if (Minecraft.getMinecraft().getResourceManager() instanceof SimpleReloadableResourceManager) {
+			SimpleReloadableResourceManager resourceManager = (SimpleReloadableResourceManager) Minecraft.getMinecraft().getResourceManager();
+			resourceManager.registerReloadListener(this);
+
+			// Forces this resource pack to be loaded early on. FML already did it's initial registration so we need to bypass that.
+			FallbackResourceManager domainManager = domainResourceManagers.get(resourceManager).get(modID);
+			resourcePacks.get(domainManager).add(this);
+		}
+	}
+
+	private static File getPackFileByModID(String modID) {
+		IResourcePack pack = FMLClientHandler.instance().getResourcePackFor(modID);
+		if (pack instanceof LegacyV2Adapter) {
+			return resourcePackFile.get(legacyPack.get(pack));
+		} else if (pack instanceof AbstractResourcePack) {
+			return resourcePackFile.get(pack);
+		}
+		return null;
+	}
+
+	public static ResourceLocation nameToLocation(String name) {
+		name = name.substring(name.indexOf("/") + 1);
+		String domain = name.substring(0, name.indexOf("/"));
+		String path = name.substring(name.indexOf("/") + 1);
+
+		LOG.debug("Converted " + name + " to " + new ResourceLocation(domain, path));
+		return new ResourceLocation(domain, path);
+	}
+
+	private static Map<Pair<RegistryDelegate<Item>, Integer>, ModelResourceLocation> getCustomModels() {
+		try {
+			Field field = ModelLoader.class.getDeclaredField("customModels");
+			return (Map<Pair<RegistryDelegate<Item>, Integer>, ModelResourceLocation>) FieldUtils.readStaticField(field, true);
+		} catch (Exception e) {
+			LOG.error("Caught exception getting customModels from the model loader, ", e);
+		}
+
+		return Maps.newHashMap();
+	}
+
+	@Override
+	protected InputStream getInputStreamByName(String name) throws IOException {
+		// Default to fallback if possible.
+		if (!((boolean) hasResourceName.invoke(realResourcePack, name))) {
+			if (cache.containsKey(name)) {
+				LOG.debug("ConcreteResourcePack was asked to obtain: " + name + " using cache.");
+				return IOUtils.toInputStream(cache.get(name));
+			}
+
+			LOG.debug("ConcreteResourcePack was asked to obtain: " + name);
+			if (isLocation(name, "/blockstates/")) {
+				return IOUtils.toInputStream(getBlockState(name));
+			} else if (isLocation(name, "/models/block/")) {
+				return IOUtils.toInputStream(getBlockModel(name));
+			} else if (isLocation(name, "/models/item/")) {
+				return IOUtils.toInputStream(getItemModel(name));
+			}
+		}
+
+		// Cover ourself in the event of something odd happening.
```

Comment:
The something odd is if we're requested to get an asset that we can't generate. Will adjust docs.